lucene/core/src/java/org/apache/lucene/codecs/lucene80/IndexedDISI.java

```diff
@@ -0,0 +1,542 @@
+        for (long word : buffer.getBits()) {
+          out.writeLong(word);
+        }
+      }
+    } else {
+      BitSetIterator it = new BitSetIterator(buffer, cardinality);
+      for (int doc = it.nextDoc(); doc != DocIdSetIterator.NO_MORE_DOCS; doc = it.nextDoc()) {
+        out.writeShort((short) doc);
+      }
+    }
+  }
+
+  // Creates a DENSE rank-entry (the number of set bits up to a given point) for the buffer.
+  // One rank-entry for every 512 bits/8 longs for a total of 128 * 16 bits.
+  // Represented as a byte[] for fast flushing and mirroring of the retrieval representation.
+  private static byte[] createRank(FixedBitSet buffer) {
+    final byte[] rank = new byte[RANKS_PER_BLOCK*2];
+    final long[] bits = buffer.getBits();
+    int bitCount = 0;
+    for (int word = 0 ; word < DENSE_BLOCK_LONGS ; word++) {
+      if ((word & 0x07) == 0) { // Every 8 longs
+        rank[word >> 2] = (byte)(bitCount>>8);
+        rank[(word >> 2)+1] = (byte)(bitCount & 0xFF);
+      }
+      bitCount += Long.bitCount(bits[word]);
+    }
+    return rank;
+  }
+
+  /**
+   * Writes the docIDs from it to out, in logical blocks, one for each 65536 docIDs in monotonically
+   * increasing gap-less order.
+   * @param it  the document IDs.
+   * @param out destination for the blocks.
+   * @throws IOException if there was an error writing to out.
+   * @return the number of jump-table entries following the blocks, -1 for no entries. This should be stored in meta.
+   */
+  static short writeBitSet(DocIdSetIterator it, IndexOutput out) throws IOException {
+    final long origo = out.getFilePointer(); // All jumps are relative to the origo
+    int totalCardinality = 0;
+    int blockCardinality = 0;
+    final FixedBitSet buffer = new FixedBitSet(1<<16);
+    long[] jumps = new long[ArrayUtil.oversize(1, Long.BYTES)];
+    jumps[0] = out.getFilePointer()-origo; // First block starts at index 0
+    int prevBlock = -1;
+    int jumpBlockIndex = 0;
+
+    for (int doc = it.nextDoc(); doc != DocIdSetIterator.NO_MORE_DOCS; doc = it.nextDoc()) {
+      final int block = doc >>> 16;
+      if (prevBlock != -1 && block != prevBlock) {
+        // Track offset+index from previous block up to current
+        jumps = addJumps(jumps, out.getFilePointer()-origo, totalCardinality, jumpBlockIndex, prevBlock+1);
+        jumpBlockIndex = prevBlock+1;
+        // Flush block
+        flush(prevBlock, buffer, blockCardinality, out);
+        // Reset for next block
+        buffer.clear(0, buffer.length());
+        totalCardinality += blockCardinality;
+        blockCardinality = 0;
+      }
+      buffer.set(doc & 0xFFFF);
+      blockCardinality++;
+      prevBlock = block;
+    }
+    if (blockCardinality > 0) {
+      jumps = addJumps(jumps, out.getFilePointer()-origo, totalCardinality, jumpBlockIndex, prevBlock+1);
+      flush(prevBlock, buffer, blockCardinality, out);
+      buffer.clear(0, buffer.length());
+      prevBlock++;
+    }
```

Comment:
The `lastBlock`handles that so that the jumpTable will have its length shortened. But it does have the unintended consequence that advancing into trailing empty blocks will not use jumpTables.

The check is `blockIndex < jumpTableEntryCount` in `IndexedDISI.advanceBlock` and I _think_ it can be optimized by seeking to the last jumpTable entry. I'll have to look at that later though.

Reply:
Ah... The `writeBitset`-method takes an `DocIDSetIterator`, which has no concept of the range of documents, only the ones which are present. It is not possible to make jump-table entries for trailing blank blocks as there are none at write time.

I have changed `IndexedDISI.advanceBlock` to handle jumps outside of the known range by jumping to the last known entry, which will always be the `NO_MORE_DOCS`-entry.